File: src/analyzers/types/ai.ts
/**
 * AI 협업 추적 관련 타입 정의
 */

/**
 * 지원하는 AI 도구
 */
export enum AITool {
  CLAUDE = 'claude',
  GITHUB_COPILOT = 'github_copilot',
  CHATGPT = 'chatgpt',
  CURSOR = 'cursor',
  TABNINE = 'tabnine',
  CODEWHISPERER = 'codewhisperer',
  OTHER = 'other'
}

/**
 * AI 도구 사용 유형
 */
export enum AIUsageType {
  CODE_GENERATION = 'code_generation',
  CODE_COMPLETION = 'code_completion',
  CODE_EXPLANATION = 'code_explanation',
  CODE_REVIEW = 'code_review',
  DEBUGGING = 'debugging',
  REFACTORING = 'refactoring',
  DOCUMENTATION = 'documentation',
  TESTING = 'testing',
  ARCHITECTURE = 'architecture',
  OTHER = 'other'
}

/**
 * AI 제안 상태
 */
export enum AISuggestionStatus {
  ACCEPTED = 'accepted',
  REJECTED = 'rejected',
  MODIFIED = 'modified',
  PENDING = 'pending'
}

/**
 * AI 사용 감지 결과
 */
export interface AIUsageDetection {
  tool: AITool;
  usageType: AIUsageType;
  timestamp: number;
  filePath?: string;
  lineRange?: {
    start: number;
    end: number;
  };
  prompt?: string;
  suggestion?: string;
  confidence: number;
  context?: string;
}

/**
 * AI 제안 추적
 */
export interface AISuggestion {
  id: string;
  tool: AITool;
  usageType: AIUsageType;
  status: AISuggestionStatus;
  timestamp: number;
  filePath: string;
  originalCode?: string;
  suggestedCode: string;
  acceptedCode?: string;
  modificationRatio?: number; // 0-1, 수정된 비율
  responseTime?: number; // ms
  tokenCount?: number;
}

/**
 * AI 도구 세션
 */
export interface AISession {
  id: string;
  tool: AITool;
  startTime: number;
  endTime?: number;
  interactions: AIInteraction[];
  totalTokens?: number;
  totalCost?: number;
}

/**
 * AI 상호작용
 */
export interface AIInteraction {
  timestamp: number;
  type: AIUsageType;
  prompt?: string;
  response?: string;
  tokenCount?: number;
  duration?: number; // ms
  result?: AISuggestionStatus;
}

/**
 * AI 효과성 메트릭
 */
export interface AIEffectivenessMetrics {
  tool: AITool;
  timeWindow: 'hour' | 'day' | 'week' | 'month';
  acceptanceRate: number; // 0-1
  modificationRate: number; // 0-1
  averageResponseTime: number; // ms
  totalInteractions: number;
  successfulInteractions: number;
  tokenEfficiency: number; // 유용한 코드 / 총 토큰
  timesSaved: number; // 예상 절약 시간 (분)
  codeQualityImpact: {
    bugsIntroduced: number;
    bugsFixed: number;
    complexityChange: number;
    testCoverageChange: number;
  };
}

/**
 * AI 사용 패턴
 */
export interface AIUsagePattern {
  tool: AITool;
  patterns: {
    mostUsedTypes: AIUsageType[];
    peakHours: number[]; // 0-23
    averageSessionDuration: number; // 분
    preferredFileTypes: string[];
    commonPromptPatterns: string[];
  };
  productivity: {
    linesGeneratedPerHour: number;
    functionsCompletedPerDay: number;
    testsWrittenPerWeek: number;
  };
}

/**
 * AI 코드 품질 분석
 */
export interface AICodeQualityAnalysis {
  suggestion: AISuggestion;
  quality: {
    readability: number; // 0-100
    maintainability: number; // 0-100
    performance: number; // 0-100
    security: number; // 0-100
    testability: number; // 0-100
  };
  issues: {
    type: 'bug' | 'vulnerability' | 'code_smell' | 'performance';
    severity: 'low' | 'medium' | 'high' | 'critical';
    description: string;
  }[];
  improvements: string[];
}

/**
 * AI 협업 분석 결과
 */
export interface AICollaborationAnalysis {
  timestamp: number;
  tools: {
    [key in AITool]?: {
      sessions: number;
      interactions: number;
      acceptanceRate: number;
      timesSaved: number;
      preferredUsageTypes: AIUsageType[];
    };
  };
  overallMetrics: {
    totalAIAssistedTime: number; // 분
    humanCodingTime: number; // 분
    aiContributionRatio: number; // 0-1
    productivityGain: number; // 백분율
  };
  insights: {
    mostEffectiveTool: AITool;
    bestUseCases: AIUsageType[];
    improvementAreas: string[];
    recommendations: string[];
  };
  trends: {
    adoptionRate: number[]; // 시간별 사용률
    qualityTrend: number[]; // 시간별 품질 점수
    efficiencyTrend: number[]; // 시간별 효율성
  };
}

/**
 * AI 모니터 이벤트
 */
export interface AIMonitorEvents {
  aiUsageDetected: (detection: AIUsageDetection) => void;
  suggestionTracked: (suggestion: AISuggestion) => void;
  sessionStarted: (session: AISession) => void;
  sessionEnded: (session: AISession) => void;
  metricsUpdated: (metrics: AIEffectivenessMetrics) => void;
  patternIdentified: (pattern: AIUsagePattern) => void;
  qualityAnalyzed: (analysis: AICodeQualityAnalysis) => void;
}

/**
 * AI 도구 설정
 */
export interface AIToolConfig {
  tool: AITool;
  enabled: boolean;
  apiEndpoint?: string;
  apiKey?: string;
  model?: string;
  maxTokens?: number;
  temperature?: number;
  detectionPatterns?: RegExp[];
  filePatterns?: string[];
}

/**
 * AI 감지 규칙
 */
export interface AIDetectionRule {
  tool: AITool;
  patterns: {
    comment?: RegExp[];
    code?: RegExp[];
    file?: RegExp[];
    process?: RegExp[];
  };
  confidence: number;
}

File: python/src/devflow_monitor/analyzers/types/ai.py
"""
AI Collaboration Tracking Type Definitions.

This module defines types for tracking AI tool usage including
Claude, GitHub Copilot, ChatGPT, Cursor, TabNine, and CodeWhisperer.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Literal

from pydantic import BaseModel, Field


class AITool(str, Enum):
    """Supported AI tools."""

    CLAUDE = "claude"
    GITHUB_COPILOT = "github_copilot"
    CHATGPT = "chatgpt"
    CURSOR = "cursor"
    TABNINE = "tabnine"
    CODEWHISPERER = "codewhisperer"
    OTHER = "other"


class AIUsageType(str, Enum):
    """Types of AI tool usage."""

    CODE_GENERATION = "code_generation"
    CODE_COMPLETION = "code_completion"
    CODE_EXPLANATION = "code_explanation"
    CODE_REVIEW = "code_review"
    DEBUGGING = "debugging"
    REFACTORING = "refactoring"
    DOCUMENTATION = "documentation"
    TESTING = "testing"
    ARCHITECTURE = "architecture"
    OTHER = "other"


class AISuggestionStatus(str, Enum):
    """Status of AI suggestions."""

    ACCEPTED = "accepted"
    REJECTED = "rejected"
    MODIFIED = "modified"
    PENDING = "pending"


class AIUsageDetection(BaseModel):
    """
    AI usage detection result.

    Records when AI tool usage is detected.

    Attributes:
        tool: The AI tool detected.
        usage_type: Type of usage.
        timestamp: Detection timestamp.
        file_path: File where detected.
        line_range: Line range if applicable.
        prompt: The prompt used (if known).
        suggestion: The suggestion made.
        confidence: Detection confidence (0-1).
        context: Additional context.
    """

    tool: AITool
    usage_type: AIUsageType
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    file_path: str | None = None
    line_range: dict[str, int] | None = None  # start, end
    prompt: str | None = None
    suggestion: str | None = None
    confidence: float = Field(ge=0.0, le=1.0)
    context: str | None = None


class AISuggestion(BaseModel):
    """
    AI suggestion tracking.

    Tracks individual AI suggestions and their outcomes.

    Attributes:
        id: Unique suggestion identifier.
        tool: AI tool that made suggestion.
        usage_type: Type of usage.
        status: Suggestion status.
        timestamp: When suggested.
        file_path: Target file.
        original_code: Code before suggestion.
        suggested_code: The suggested code.
        accepted_code: Code actually used.
        modification_ratio: How much was modified (0-1).
        response_time: Response time in ms.
        token_count: Token count if known.
    """

    id: str
    tool: AITool
    usage_type: AIUsageType
    status: AISuggestionStatus
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    file_path: str
    original_code: str | None = None
    suggested_code: str
    accepted_code: str | None = None
    modification_ratio: float | None = None  # 0-1
    response_time: int | None = None  # milliseconds
    token_count: int | None = None


class AIInteraction(BaseModel):
    """
    Single AI interaction.

    Records a single interaction with an AI tool.

    Attributes:
        timestamp: When the interaction occurred.
        type: Type of interaction.
        prompt: The prompt used.
        response: The AI response.
        token_count: Token count.
        duration: Duration in ms.
        result: Outcome status.
    """

    timestamp: datetime = Field(default_factory=datetime.utcnow)
    type: AIUsageType
    prompt: str | None = None
    response: str | None = None
    token_count: int | None = None
    duration: int | None = None  # milliseconds
    result: AISuggestionStatus | None = None


class AISession(BaseModel):
    """
    AI tool session.

    Tracks a session of AI tool usage.

    Attributes:
        id: Session identifier.
        tool: AI tool used.
        start_time: Session start.
        end_time: Session end.
        interactions: List of interactions.
        total_tokens: Total tokens used.
        total_cost: Estimated cost.
    """

    id: str
    tool: AITool
    start_time: datetime = Field(default_factory=datetime.utcnow)
    end_time: datetime | None = None
    interactions: list[AIInteraction] = Field(default_factory=list)
    total_tokens: int | None = None
    total_cost: float | None = None


class AIEffectivenessMetrics(BaseModel):
    """
    AI tool effectiveness metrics.

    Measures how effective an AI tool is.

    Attributes:
        tool: The AI tool.
        time_window: Measurement time window.
        acceptance_rate: Rate of accepted suggestions (0-1).
        modification_rate: Rate of modified suggestions (0-1).
        average_response_time: Average response time in ms.
        total_interactions: Total number of interactions.
        successful_interactions: Number of successful interactions.
        token_efficiency: Useful code per token.
        times_saved: Estimated time saved in minutes.
        code_quality_impact: Impact on code quality.
    """

    tool: AITool
    time_window: Literal["hour", "day", "week", "month"] = "hour"
    acceptance_rate: float = 0.0  # 0-1
    modification_rate: float = 0.0  # 0-1
    average_response_time: int = 0  # milliseconds
    total_interactions: int = 0
    successful_interactions: int = 0
    token_efficiency: float = 0.0
    times_saved: int = 0  # minutes
    code_quality_impact: dict[str, Any] = Field(default_factory=dict)


class AIUsagePattern(BaseModel):
    """
    AI usage pattern analysis.

    Analyzes patterns in AI tool usage.

    Attributes:
        tool: The AI tool.
        patterns: Usage patterns.
        productivity: Productivity metrics.
    """

    tool: AITool
    patterns: dict[str, Any] = Field(default_factory=dict)
    productivity: dict[str, Any] = Field(default_factory=dict)


class AICodeQualityMetrics(BaseModel):
    """Code quality metrics for AI suggestions."""

    readability: int = Field(ge=0, le=100, default=0)
    maintainability: int = Field(ge=0, le=100, default=0)
    performance: int = Field(ge=0, le=100, default=0)
    security: int = Field(ge=0, le=100, default=0)
    testability: int = Field(ge=0, le=100, default=0)


class AICodeQualityIssue(BaseModel):
    """Code quality issue in AI suggestion."""

    type: Literal["bug", "vulnerability", "code_smell", "performance"]
    severity: Literal["low", "medium", "high", "critical"]
    description: str


class AICodeQualityAnalysis(BaseModel):
    """
    AI code quality analysis.

    Analyzes the quality of AI-generated code.

    Attributes:
        suggestion: The analyzed suggestion.
        quality: Quality metrics.
        issues: Detected issues.
        improvements: Suggested improvements.
    """

    suggestion: AISuggestion
    quality: AICodeQualityMetrics
    issues: list[AICodeQualityIssue] = Field(default_factory=list)
    improvements: list[str] = Field(default_factory=list)


class AICollaborationAnalysis(BaseModel):
    """
    Complete AI collaboration analysis.

    Comprehensive analysis of AI tool usage.

    Attributes:
        timestamp: Analysis timestamp.
        tools: Per-tool analysis.
        overall_metrics: Overall metrics.
        insights: Generated insights.
        trends: Usage trends.
    """

    timestamp: datetime = Field(default_factory=datetime.utcnow)
    tools: dict[AITool, dict[str, Any]] = Field(default_factory=dict)
    overall_metrics: dict[str, Any] = Field(default_factory=dict)
    insights: dict[str, Any] = Field(default_factory=dict)
    trends: dict[str, Any] = Field(default_factory=dict)


class AIDetectionRule(BaseModel):
    """
    AI detection rule.

    Defines rules for detecting AI tool usage.

    Attributes:
        tool: The AI tool to detect.
        patterns: Detection patterns by type.
        confidence: Base confidence level.
    """

    tool: AITool
    patterns: dict[str, list[str]] = Field(default_factory=dict)
    confidence: float = Field(ge=0.0, le=1.0, default=0.7)


class AIToolConfig(BaseModel):
    """
    AI tool configuration.

    Configuration for AI tool integration.

    Attributes:
        tool: The AI tool.
        enabled: Whether enabled.
        api_endpoint: API endpoint URL.
        api_key: API key (if needed).
        model: Model to use.
        max_tokens: Maximum tokens.
        temperature: Generation temperature.
        detection_patterns: Patterns for detection.
        file_patterns: File patterns to monitor.
    """

    tool: AITool
    enabled: bool = True
    api_endpoint: str | None = None
    api_key: str | None = None
    model: str | None = None
    max_tokens: int | None = None
    temperature: float | None = None
    detection_patterns: list[str] | None = None
    file_patterns: list[str] | None = None
